File: src/apps/cppbench/benchs/btreap.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/col/Treap.h>
#include <base/util/Profile.h>
#include <base/Panic.h>

#include <m3/stream/Standard.h>

#include "../cppbench.h"

using namespace m3;

struct MyTItem : public TreapNode<MyTItem, uint32_t> {
    explicit MyTItem(uint32_t _val) : TreapNode(_val), val(_val) {
    }

    uint32_t val;
};

NOINLINE static void insert() {
    struct TreapInsertRunner : public Runner {
        void run() override {
            for(uint32_t i = 0; i < 100; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void post() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(30);
    TreapInsertRunner runner;
    cout << "100-elements: " << pr.runner_with_id(runner, 0x03) << "\n";
}

NOINLINE static void find() {
    struct TreapSearchRunner : public Runner {
        void pre() override {
            for(uint32_t i = 0; i < 100; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void run() override {
            for(uint32_t i = 0; i < 100; ++i) {
                MyTItem *item = treap.find(i);
                if(!item || item->val != i)
                    PANIC("Test failed: " << (item ? item->val : 0) << " != " << i);
            }
        }
        void post() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(30);
    TreapSearchRunner runner;
    cout << "100-elements: " << pr.runner_with_id(runner, 0x02) << "\n";
}

NOINLINE static void clear() {
    struct TreapClearRunner : public Runner {
        void pre() override {
            for(uint32_t i = 0; i < 100; ++i) {
                treap.insert(new MyTItem(i));
            }
        }
        void run() override {
            MyTItem *item;
            while((item = treap.remove_root()) != nullptr) {
                delete item;
            }
        }

        Treap<MyTItem> treap;
    };

    Profile pr(30);
    TreapClearRunner runner;
    cout << "100-elements: " << pr.runner_with_id(runner, 0x01) << "\n";
}

void btreap() {
    RUN_BENCH(insert);
    RUN_BENCH(find);
    RUN_BENCH(clear);
}

File: src/apps/rust/benchs/src/btreap.rs
use m3::col::Treap;
use m3::profile;
use m3::test;

pub fn run(t: &mut test::Tester) {
    run_test!(t, insert);
    run_test!(t, find);
    run_test!(t, clear);
}

fn insert() {
    let mut prof = profile::Profiler::new().repeats(30);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl profile::Runner for BTreeTester {
        fn pre(&mut self) {
            self.0.clear();
        }
        fn run(&mut self) {
            for i in 0..100 {
                self.0.insert(i, i);
            }
        }
    }

    println!("Inserting 100 elements: {}", prof.runner_with_id(&mut BTreeTester::default(), 0x71));
}

fn find() {
    let mut prof = profile::Profiler::new().repeats(30);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl profile::Runner for BTreeTester {
        fn pre(&mut self) {
            for i in 0..100 {
                self.0.insert(i, i);
            }
        }
        fn run(&mut self) {
            for i in 0..100 {
                let val = self.0.get(&i);
                assert_eq!(val, Some(&i));
            }
        }
        fn post(&mut self) {
            self.0.clear();
        }
    }

    println!("Searching for 100 elements: {}", prof.runner_with_id(&mut BTreeTester::default(), 0x72));
}

fn clear() {
    let mut prof = profile::Profiler::new().repeats(30);

    #[derive(Default)]
    struct BTreeTester(Treap<u32, u32>);

    impl profile::Runner for BTreeTester {
        fn pre(&mut self) {
            for i in 0..100 {
                self.0.insert(i, i);
            }
        }
        fn run(&mut self) {
            self.0.clear();
        }
    }

    println!("Removing 100-element list: {}", prof.runner_with_id(&mut BTreeTester::default(), 0x73));
}
